library/src/main/java/org/xmtp/android/library/Client.kt

```diff
@@ -0,0 +1,196 @@
+        val res = apiClient.query(topics = topics)
+        for (envelope in res.envelopesList) {
+            try {
+                val encryptedBundle = EncryptedPrivateKeyBundle.parseFrom(envelope.message)
+                val bundle = encryptedBundle.decrypted(account)
+                return bundle.v1
+            } catch (e: Throwable) {
+                print("Error decoding encrypted private key bundle: $e")
+                continue
+            }
+        }
+        return null
+    }
+
+    suspend fun publishUserContact(legacy: Boolean = false) {
+        val envelopes: MutableList<MessageApiOuterClass.Envelope> = mutableListOf()
+        if (legacy) {
+            val contactBundle = ContactBundle.newBuilder().also {
+                it.v1Builder.keyBundle = privateKeyBundleV1?.toPublicKeyBundle()
+            }.build()
+
+            val envelope = MessageApiOuterClass.Envelope.newBuilder().apply {
+                address?.let {
+                    contentTopic = Topic.contact(it).description
+                }
+                timestampNs = System.currentTimeMillis() * 1_000_000
+                message = contactBundle.toByteString()
+            }.build()
+
+            envelopes.plus(envelope)
+        }
+        val contactBundle = ContactBundle.newBuilder().also {
+            it.v2Builder.keyBundle = keys?.getPublicKeyBundle()
+        }.build()
+        contactBundle.v2.keyBundle.identityKey.signature.ensureWalletSignature()
+        val envelope = MessageApiOuterClass.Envelope.newBuilder().apply {
+            address?.let {
+                contentTopic = Topic.contact(it).description
+            }
+            timestampNs = System.currentTimeMillis() * 1_000_000
+            message = contactBundle.toByteString()
+        }.build()
+        envelopes.add(envelope)
+        publish(envelopes = envelopes)
+    }
+
+    fun getUserContact(peerAddress: String): ContactBundle? {
+        return contacts.find(peerAddress)
+    }
+
+    suspend fun query(topics: List<Topic>): QueryResponse {
+        return apiClient.query(topics = topics)
+    }
+
+    suspend fun publish(envelopes: List<Envelope>): PublishResponse {
+        privateKeyBundleV1?.let {
+            address?.let { address ->
+                val authorized = AuthorizedIdentity(
+                    address = address,
+                    authorized = it.identityKey.publicKey,
+                    identity = it.identityKey
+                )
+                val authToken = authorized.createAuthToken()
+                apiClient.setAuthToken(authToken)
+            }
+        }
+
+        return apiClient.publish(envelopes = envelopes)
+    }
+
+    suspend fun ensureUserContactPublished() {
+        address?.let {
+            val contact = getUserContact(peerAddress = it)
+            if (contact != null && keys?.getPublicKeyBundle() == contact.v2.keyBundle) {
+                return
+            }
+        }
+
+        publishUserContact(legacy = true)
```

Comment:
Any reason for the legacy contact here? Could we create a v2 instead?

Reply:
This is still what we do in JS. That's going to change soon-ish, I believe.